src/components/side-by-side-rendered-diff.jsx

```diff
@@ -66,37 +52,124 @@ export default class SideBySideRenderedDiff extends React.Component {
+  const elementToRemove = viewType === 'additions' ? 'del' : 'ins';
+
+  // Remove <ins/del> in <head> before parsing; parsing will throw them out
+  // but keep their contents. That could leave us with <link> or <script>
+  // elements that should have been removed.
+  let newSource = source.replace(/<head[^]*<\/head>/i, head => {
+    return head.replace(
+      new RegExp(`<${elementToRemove}[^>]*>[^]*?</${elementToRemove}>`, 'ig'),
+      '');
+  });
+
+  const parser = new DOMParser();
+  const newDocument = parser.parseFromString(newSource, 'text/html');
+  removeChangeElements(elementToRemove, newDocument);
+  renderableDocument(newDocument, page);
+
+  const prefix = source.match(/^[^]*?<html/ig)[0];
+  newSource = prefix + newDocument.documentElement.outerHTML.slice(5);
+
+  return newSource;
+}
+
+/**
+ * Process HTML document so that it renders nicely. This includes things like
  * adding a `<base>` tag so subresources are properly fetched.
  *
- * @param {string} source
+ * @param {HTMLDocument} sourceDocument
  * @param {Page} page
- * @returns {string}
+ * @returns {HTMLDocument}
  */
-function processSource (source, page) {
+function renderableDocument (sourceDocument, page) {
+  const base = sourceDocument.createElement('base');
+  base.href = page.url;
   // <meta charset> tags don't work unless they are first, so if one is
   // present, modify <head> content *after* it.
-  const hasCharsetTag = /<meta charset[^>]+>/.test(source);
-  const headMatcher = hasCharsetTag ? /<meta charset[^>]+>/ : /<head[^>]*>/;
-  const result = source.replace(headMatcher, followTag => {
-    return `${followTag}\n<base href="${page.url}">\n`;
+  const charsetElement = sourceDocument.querySelector('meta[charset]');
+  if (charsetElement) {
+    charsetElement.insertAdjacentElement('afterend', base);
+  }
+  else {
+    sourceDocument.head.insertAdjacentElement('afterbegin', base);
+  }
+
+  // The differ currently HTML-encodes the source code in these elements :\
+  // https://github.com/edgi-govdata-archiving/web-monitoring-processing/issues/94
+  sourceDocument.querySelectorAll('style, script').forEach(element => {
+    element.textContent = element.textContent
+      .replace(/&amp;/g, '&')
+      .replace(/&lt;/g, '<')
+      .replace(/&gt;/g, '>')
+      .replace(/&quot;/g, '"')
+      .replace(/&#(x?)([0-9a-f]+);/ig, (text, hex, value) => {
+        const code = parseInt(value, hex ? 16 : 10);
+        return String.fromCharCode(code);
+      });
   });
 
-  return result;
+  return sourceDocument;
+}
+
+/**
+ * Remove HTML elements representing additions or removals from a document.
+ * If removing an element leaves its parent element empty, the parent element
+ * is also removed, and so on recursively up the tree. This is meant to
+ * compensate for the fact that our diff is really a text diff that is
+ * sensitive to the tree and not an actual tree diff.
+ *
+ * NOTE: this method is meant to be converted to source code and run *in the
```

Comment:
Just curious, what is meant by this and why we default to using window.document if no sourceDocument is supplied?

Reply:
Argh, I should have removed that comment; it’s no longer meant to do that. It makes more sense in the context of https://github.com/edgi-govdata-archiving/web-monitoring-ui/pull/121/commits/a8d7cd4df8fa935c05f5773ea9482479bcb41bd6 (see line 90).